Thesis deposit

murny opened the pull request:
#709 

TODO:
- Refactor views to use partials -- might be tough, lots of differences like urls, i18n, some are hugely different etc. Leaving this off the table for now

murny commented on app/models/concerns/draft_properties.rb:
```diff
@@ -0,0 +1,133 @@
+module DraftProperties
```
Moved as much as possible between draft items and theses models into a shared concern

murny commented on app/controllers/concerns/draft_actions.rb:
```diff
@@ -0,0 +1,146 @@
+module DraftActions
```
Moved as much as possible for the draft actions for draft controller for both theses and items into a shared concern. Also did same for the file actions controller

murny commented on app/views/items/draft/describe_item.html.erb:
```diff
@@ -1,4 +1,4 @@
-<% page_title(header) %>
+<% page_title(t('.header')) %>
```
This fixes a bug in production right now. Currently the general title is repeated twice...for example if im on deposit item, describe item step, the page title becomes:

`Deposit Item | Deposit Item | ERA`

When it should be

`Describe Item | Deposit Item | ERA`

murny commented on app/views/collections/show.html.erb:
```diff
@@ -37,11 +37,22 @@
-        <%= fa_icon('plus-circle') %>
-        <%= t('.deposit_item') %>
+
+    <% if @collection.restricted.blank? %>
```
Only show deposit button on collection show page if you can actually deposit inside this community/collection pair. 

Previously, you could go to the FGSR thesis collection and press the deposit item button, which once you fill out the describe form, you will get an error saying this collection is restricted. Which doesn't make much sense.

pgwillia commented on db/seeds.rb:
```diff
@@ -21,7 +21,7 @@
   # start fresh
   [Announcement, ActiveStorage::Blob, ActiveStorage::Attachment, JupiterCore::AttachmentShim,
-   Identity, User, Type, Language].each(&:destroy_all)
+   Identity, User, Type, Language, Institution].each(&:destroy_all)
```
Ah, this is why it wasn't working for me without re-seeding data.

murny replied:
More likely was the Line on 413-416. Where you need Institutions seeded otherwise nothing shows up in the form

pgwillia replied:
I also meant to say nice refactor for the item/thesis drafts.

cwant commented on app/controllers/admin/theses/draft_controller.rb:
```diff
@@ -0,0 +1,42 @@
+  private
+
+  def authorize?
+    false
```
What does `false` here mean -- the item is open to everybody and doesn't go through Pundit?

murny replied:
Because thesis deposit is nested under /admin, it inherits the admin authorization:
https://github.com/ualbertalib/jupiter/blob/master/app/controllers/admin/admin_controller.rb#L10

(its protected also on the route level via a constraint)

So we don't need to do any further authorization checking within each action here. Hence we just turn it off for theses. 

If that makes sense?

cwant replied:
Does make sense. As somebody who hasn't looked at the stuff in a while, that wasn't obvious. Just a suggestion, but it might not hurt to add a comment, or rename to `already_authorized_via_admin?` or such (flipping the true/false).

murny replied:
Good suggestion, can rename to make it better

cwant commented on app/models/thesis.rb:
```diff
@@ -55,6 +55,76 @@ def item_type_with_status_code
+    end
+
+    draft_thesis.uuid = thesis.id
+    draft_thesis.save!
```
Is there a reason to keep the draft around? (In case they want to edit again?) Looks like this is called when the draft is converted into a Fedora object and the user is redirected to the show page.

murny replied:
Yeah its for editing again. We add the Thesis's ID to the Draft Thesis uuid column and they are kept in sync like this. Then when a user edits a Item/Thesis, we look up to see if a draft version exists? If so use that otherwise create a new draft object.

(Also future proofs us a bit, maybe one day we get rid of Fedora's Thesis's and Draft Thesis's become the new Thesis object)

cwant commented on app/controllers/concerns/draft_actions.rb:
```diff
@@ -0,0 +1,146 @@
+    before_action :set_draft, only: [:show, :update, :destroy]
+  end
+
+  def show
```
Seems that the `show` action is really an `edit` action. This is fine, but I wonder if it should have been called `edit` rather than `show`.

murny replied:
I'd agree but this comes from the wicked gem that we are using for the wizard steps and how they like it: https://github.com/schneems/wicked

No documentation on changing it to edit and probably wouldn't work. Would make urls a bit weird too (`localhost:3000/admin/theses/1/draft/describe_item/edit`, with `edit` at the end of everything)

cwant commented on app/controllers/concerns/draft_actions.rb:
```diff
@@ -0,0 +1,146 @@
+  end
+
+  def item_class
+    Item
```
It's pretty cool how you're able to do this, thanks to the abstract code in the concern. One thought: potentially you could move more code to the concern by deriving these private methods from `item_class` (e.g., construct the needed symbols using `item_class.name.underscore` or `item_class.model_name.singular`)

murny commented:
Replying to https://github.com/ualbertalib/jupiter/pull/761#pullrequestreview-138861597

If you edit a thesis which belongs to a non restricted collection (AKA not FGSR or St. Stephen's), you will be forced to update the community/collection pair and choose either FGSR/St Stephens. This is why you got an error and couldn't proceed past the describe thesis step. 

In other environments we will need to seed these values. No different then what we do for other dropdowns like Types, etc. Currently they are static and probably won't change. But who knows?

Testing can be improved in general. Some of these I hardly touched at all and show general gaps in our testing (item controller has no controller test at all and stub file was recently deleted: https://github.com/ualbertalib/jupiter/pull/759/files/e289a2f3355f0506ba6d99b66c2f898e8ce2d0d7#r202116700) . Model tests could be improved, but its a trade off...these things that are missing are things touching fedora directly which are finicky to break and very slow. Plus most of these gaps are covered by system tests. Were still at 88% + coverage for `rails test` and 82% + for `rails test:system`. If these drop below 80% then absolutely we should revisit testing and add some more tests. Can also bring back coveralls (https://coveralls.io/) and fail builds to enforce this like we do on pushmi_pullyu

cwant commented on app/models/draft_thesis.rb:
```diff
@@ -0,0 +1,167 @@
+
+  include DraftProperties
+
+  DEFAULT_RIGHTS = 'This thesis is made available by the University of Alberta Libraries'\
```
Should this go through `I18n`?

murny replied:
Ooh good find. Absolutely should go through i18n

murny replied:
This is done now. Plus the graduation term dropdown has also been `i18n`'d

cwant commented on app/views/admin/theses/draft/_progress_bar.html.erb:
```diff
@@ -0,0 +1,28 @@
+<% page_title(header) %>
```
Should setting the page title happen in this partial? Don't know if this is related to the bug you mentioned elsewhere in the conversation about this PR.

murny replied:
Basically how the title helper works is its a stack, where the titles get push at the end of the stack, aka (first in, first out). So you push titles to it, it creates an array [ 'decribe item', 'deposit item' , 'era'] and then it creates the `describe item | deposit item | era` for the page title.

So the step partials will add it's page title piece (aka the "Describe Thesis"/"Review and deposit thesis"/"Choose licence and visibility", ETC), and the H2 title with the step title.

This progress bar partial will add the generic page title, add the H1 with the generic page title and the progress bar.

(bug was the step partials were adding the generic title, instead of it's step title)

cwant commented on test/controllers/admin/theses/draft_controller_test.rb:
```diff
@@ -0,0 +1,266 @@
+    assert_response :success
+  end
+
+  test 'should prevent prevent user from skipping ahead with uncompleted steps' do
```
Sounds like you really want to 'prevent' this :)

cwant commented on test/controllers/items/files_controller_test.rb:
```diff
@@ -27,17 +27,14 @@ def before_all
-    assert_difference('ActiveStorage::Attachment.count', 1) do
-      post item_files_url(@draft_item), params: { file: @file_attachment }, xhr: true
-    end
+    # TODO: Why this failing?
```
Thought: related to file ingest now getting background-ed by the recent ActiveStorage work?

murny replied:
You might be onto something :thinking:

cwant commented on test/models/draft_thesis_test.rb:
```diff
@@ -0,0 +1,214 @@
+      rights: 'License text goes here'
+    )
+
+    assert_not draft_thesis.valid?
```
Other tests test the error message, but not this one -- don't know if that was intentional

murny replied:
Fix these, adding the extra assertion for specific error message or a number of errors coming back

murny commented:
Hey @cwant or anyone, can you please give this another review when you get the chance?

- Fixed feedback from review:
  - Add better testing of error messages
  - fix double "prevent"
  - I18n a few missing things
  - renamed needs authorization method and add comment
  - derived a bunch of methods within concerns using `draft_class` and `item_class` instead
- Added Draft Theses to cleanup rake task
- Added Draft Theses to show under user profile or users show page
